File: products/windshift/internal/routes/pages.go
package routes

import "net/http"

// RegisterPageRoutes registers workspace knowledge-pages endpoints. All
// routes are workspace-scoped; permission failures inside the handlers
// return 404 (memory: workspace-resource access checks must not leak
// existence).
func RegisterPageRoutes(deps *Deps) {
	if deps.Pages.Page == nil {
		return
	}
	api := deps.API
	auth := deps.AuthMiddleware.RequireAuth

	api.HandleH("GET /workspaces/{workspaceId}/pages/tree", auth(http.HandlerFunc(deps.Pages.Page.GetTree)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/archived", auth(http.HandlerFunc(deps.Pages.Page.ListArchived)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/search", auth(http.HandlerFunc(deps.Pages.Page.Search)))
	api.HandleH("POST /workspaces/{workspaceId}/pages", auth(http.HandlerFunc(deps.Pages.Page.Create)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/{pageId}", auth(http.HandlerFunc(deps.Pages.Page.Get)))
	api.HandleH("POST /workspaces/{workspaceId}/pages/{pageId}/unarchive", auth(http.HandlerFunc(deps.Pages.Page.Unarchive)))
	api.HandleH("PUT /workspaces/{workspaceId}/pages/{pageId}", auth(http.HandlerFunc(deps.Pages.Page.Update)))
	api.HandleH("DELETE /workspaces/{workspaceId}/pages/{pageId}", auth(http.HandlerFunc(deps.Pages.Page.Delete)))
	api.HandleH("POST /workspaces/{workspaceId}/pages/{pageId}/move", auth(http.HandlerFunc(deps.Pages.Page.Move)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/{pageId}/history", auth(http.HandlerFunc(deps.Pages.Page.GetHistory)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/{pageId}/history/{revisionId}", auth(http.HandlerFunc(deps.Pages.Page.GetRevision)))
	api.HandleH("POST /workspaces/{workspaceId}/pages/{pageId}/history/{revisionId}/restore", auth(http.HandlerFunc(deps.Pages.Page.RestoreRevision)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/{pageId}/permissions", auth(http.HandlerFunc(deps.Pages.Page.GetPermissions)))
	api.HandleH("POST /workspaces/{workspaceId}/pages/{pageId}/permissions", auth(http.HandlerFunc(deps.Pages.Page.GrantPermission)))
	api.HandleH("DELETE /workspaces/{workspaceId}/pages/{pageId}/permissions/{permissionId}", auth(http.HandlerFunc(deps.Pages.Page.RevokePermission)))
	api.HandleH("PATCH /workspaces/{workspaceId}/pages/{pageId}/inheritance", auth(http.HandlerFunc(deps.Pages.Page.SetInheritance)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/{pageId}/diagrams", auth(http.HandlerFunc(deps.Pages.Page.ListDiagrams)))
	api.HandleH("POST /workspaces/{workspaceId}/pages/{pageId}/diagrams", auth(http.HandlerFunc(deps.Pages.Page.CreateDiagram)))
	api.HandleH("GET /workspaces/{workspaceId}/pages/{pageId}/diagrams/{attachmentId}", auth(http.HandlerFunc(deps.Pages.Page.GetDiagram)))
	api.HandleH("PUT /workspaces/{workspaceId}/pages/{pageId}/diagrams/{attachmentId}", auth(http.HandlerFunc(deps.Pages.Page.UpdateDiagram)))

	if deps.Pages.KnowledgeSearch != nil {
		api.HandleH("GET /workspaces/{workspaceId}/knowledge/search", auth(http.HandlerFunc(deps.Pages.KnowledgeSearch.Search)))
	}

	if deps.Pages.PageLabel != nil {
		// Workspace-scoped label CRUD. Permission failures inside the
		// handler return 404 (memory: workspace-resource access checks
		// must not leak existence).
		api.HandleH("GET /workspaces/{workspaceId}/page-labels", auth(http.HandlerFunc(deps.Pages.PageLabel.List)))
		api.HandleH("POST /workspaces/{workspaceId}/page-labels", auth(http.HandlerFunc(deps.Pages.PageLabel.Create)))
		api.HandleH("GET /workspaces/{workspaceId}/page-labels/{labelId}", auth(http.HandlerFunc(deps.Pages.PageLabel.Get)))
		api.HandleH("PUT /workspaces/{workspaceId}/page-labels/{labelId}", auth(http.HandlerFunc(deps.Pages.PageLabel.Update)))
		api.HandleH("DELETE /workspaces/{workspaceId}/page-labels/{labelId}", auth(http.HandlerFunc(deps.Pages.PageLabel.Delete)))

		// Page-scoped attachments — gated per-page via PagePermissionService.
		api.HandleH("GET /workspaces/{workspaceId}/pages/{pageId}/labels", auth(http.HandlerFunc(deps.Pages.PageLabel.ListForPage)))
		api.HandleH("PUT /workspaces/{workspaceId}/pages/{pageId}/labels", auth(http.HandlerFunc(deps.Pages.PageLabel.SetForPage)))
		api.HandleH("POST /workspaces/{workspaceId}/pages/{pageId}/labels", auth(http.HandlerFunc(deps.Pages.PageLabel.AddToPage)))
		api.HandleH("DELETE /workspaces/{workspaceId}/pages/{pageId}/labels/{labelId}", auth(http.HandlerFunc(deps.Pages.PageLabel.RemoveFromPage)))
	}
}

File: products/windshift/frontend/src/lib/api/pages.js
import { fetchAPI } from './core.js';
import { buildQueryString } from './utils.js';

/**
 * Workspace knowledge-pages API client. Mirrors logbook.js style: every
 * method returns a Promise from fetchAPI; auth is automatic via cookies
 * (core.js sets credentials: 'same-origin').
 */
export const pages = {
  /** Fetch the workspace page tree + flat list. */
  getTree: (workspaceId) => fetchAPI(`/workspaces/${workspaceId}/pages/tree`),

  /** Fetch a single page (404 on missing or no view permission). */
  getPage: (workspaceId, pageId) => fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}`),

  /** Create a new page. parentId is optional (null/undefined = root). */
  createPage: (
    workspaceId,
    { title, content = '', parentId = null, isHome = false, metadata = {} }
  ) =>
    fetchAPI(`/workspaces/${workspaceId}/pages`, {
      method: 'POST',
      body: JSON.stringify({ title, content, parent_id: parentId, is_home: isHome, metadata }),
    }),

  /**
   * Update title/content on a page. Inheritance has its own admin-gated
   * endpoint (setInheritance below) — do not send the flag here; the
   * server rejects it as an unknown field shape and an editor without
   * admin would otherwise be able to flip inheritance via a normal save.
   */
  updatePage: (
    workspaceId,
    pageId,
    { title, content, metadata = undefined, expectedContentHash = undefined }
  ) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}`, {
      method: 'PUT',
      body: JSON.stringify({
        title,
        content,
        ...(metadata === undefined ? {} : { metadata }),
        ...(expectedContentHash === undefined
          ? {}
          : { expected_content_hash: expectedContentHash }),
      }),
    }),

  /** Archive a page (and every descendant). */
  archivePage: (workspaceId, pageId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}`, { method: 'DELETE' }),

  /** Admin-only: list every archived page in the workspace with archiver display name. */
  listArchived: (workspaceId) => fetchAPI(`/workspaces/${workspaceId}/pages/archived`),

  /** Admin-only: clear archived_at/archived_by on a single page (no content overwrite). */
  unarchive: (workspaceId, pageId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/unarchive`, { method: 'POST' }),

  /**
   * Reparent a page; pass parentId=null to move it to the workspace root.
   * prevSiblingId / nextSiblingId position the page within its new parent's
   * children — either may be null for "start of list" / "end of list", and
   * omitting both preserves the legacy append-by-natural-order behavior.
   */
  movePage: (
    workspaceId,
    pageId,
    parentId,
    { destinationWorkspaceId = null, prevSiblingId = null, nextSiblingId = null } = {}
  ) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/move`, {
      method: 'POST',
      body: JSON.stringify({
        ...(destinationWorkspaceId != null
          ? { destination_workspace_id: destinationWorkspaceId }
          : {}),
        parent_id: parentId,
        prev_sibling_id: prevSiblingId,
        next_sibling_id: nextSiblingId,
      }),
    }),

  /** Paginated revision history for a page. */
  getHistory: (workspaceId, pageId, { limit = 50, offset = 0 } = {}) =>
    fetchAPI(
      `/workspaces/${workspaceId}/pages/${pageId}/history${buildQueryString({ limit, offset })}`
    ),

  /** Fetch a single revision; must belong to the page. */
  getRevision: (workspaceId, pageId, revisionId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/history/${revisionId}`),

  /** Restore a revision; produces a new revision of type 'restore'. */
  restoreRevision: (workspaceId, pageId, revisionId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/history/${revisionId}/restore`, {
      method: 'POST',
    }),

  /** Create an immutable diagram attachment and atomically insert its Page fence. */
  createDiagram: (
    workspaceId,
    pageId,
    {
      name,
      mermaid = undefined,
      excalidraw = undefined,
      placement = 'end',
      expectedContentHash = undefined,
    }
  ) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/diagrams`, {
      method: 'POST',
      body: JSON.stringify({
        name,
        ...(mermaid ? { mermaid } : {}),
        ...(excalidraw ? { excalidraw } : {}),
        placement,
        ...(expectedContentHash ? { expected_content_hash: expectedContentHash } : {}),
      }),
    }),

  /** Replace one Page diagram through the shared immutable lifecycle. */
  updateDiagram: (
    workspaceId,
    pageId,
    attachmentId,
    {
      name = undefined,
      mermaid = undefined,
      excalidraw = undefined,
      expectedContentHash = undefined,
    }
  ) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/diagrams/${attachmentId}`, {
      method: 'PUT',
      body: JSON.stringify({
        ...(name ? { name } : {}),
        ...(mermaid ? { mermaid } : {}),
        ...(excalidraw ? { excalidraw } : {}),
        ...(expectedContentHash ? { expected_content_hash: expectedContentHash } : {}),
      }),
    }),

  /** Read-only effective permissions + own ACL rows. */
  getPermissions: (workspaceId, pageId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/permissions`),

  /** Grant a new ACL row on a page. Requires page.admin on the target. */
  grantPermission: (workspaceId, pageId, { principalType, principalId, permissionLevel }) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/permissions`, {
      method: 'POST',
      body: JSON.stringify({
        principal_type: principalType,
        principal_id: principalId,
        permission_level: permissionLevel,
      }),
    }),

  /** Revoke a single ACL row. The row must belong to the named page. */
  revokePermission: (workspaceId, pageId, permissionId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/permissions/${permissionId}`, {
      method: 'DELETE',
    }),

  /** Toggle the inherit_permissions flag on a page. */
  setInheritance: (workspaceId, pageId, inheritPermissions) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/inheritance`, {
      method: 'PATCH',
      body: JSON.stringify({ inherit_permissions: inheritPermissions }),
    }),

  /** Unified knowledge search across pages (and future sources). */
  searchKnowledge: (workspaceId, query, { limit = 25 } = {}) =>
    fetchAPI(`/workspaces/${workspaceId}/knowledge/search${buildQueryString({ q: query, limit })}`),

  /**
   * Title-substring page search scoped to a workspace. Server-side and
   * permission-filtered. Used by the page picker in the link dialog and
   * the page-side work-item popover.
   */
  searchPages: (workspaceId, query, { limit = 20 } = {}) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/search${buildQueryString({ q: query, limit })}`),
};

/**
 * Page label API client. Workspace-scoped labels that attach to pages only;
 * separate system from work-item labels (no shared rows or endpoints).
 *
 * Label CRUD requires workspace `page.edit`; attach/detach requires
 * per-page edit (evaluated by PagePermissionService). Permission failures
 * surface as 404 to avoid leaking page-label existence.
 */
export const pageLabels = {
  list: (workspaceId) => fetchAPI(`/workspaces/${workspaceId}/page-labels`),
  get: (workspaceId, id) => fetchAPI(`/workspaces/${workspaceId}/page-labels/${id}`),
  create: (workspaceId, { name, color }) =>
    fetchAPI(`/workspaces/${workspaceId}/page-labels`, {
      method: 'POST',
      body: JSON.stringify({ name, color }),
    }),
  update: (workspaceId, id, { name, color }) =>
    fetchAPI(`/workspaces/${workspaceId}/page-labels/${id}`, {
      method: 'PUT',
      body: JSON.stringify({ name, color }),
    }),
  delete: (workspaceId, id) =>
    fetchAPI(`/workspaces/${workspaceId}/page-labels/${id}`, { method: 'DELETE' }),

  listForPage: (workspaceId, pageId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/labels`),
  setForPage: (workspaceId, pageId, labelIds) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/labels`, {
      method: 'PUT',
      body: JSON.stringify({ label_ids: labelIds }),
    }),
  addToPage: (workspaceId, pageId, labelId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/labels`, {
      method: 'POST',
      body: JSON.stringify({ label_id: labelId }),
    }),
  removeFromPage: (workspaceId, pageId, labelId) =>
    fetchAPI(`/workspaces/${workspaceId}/pages/${pageId}/labels/${labelId}`, {
      method: 'DELETE',
    }),
};
